arch/cortex-m4/src/lib.rs

```diff
@@ -202,148 +346,11 @@ pub unsafe extern "C" fn hard_fault_handler() {
-             \tbottom of stack  0x{:x}\r\n\
-             \tSHCSR 0x{:x}\r\n\
-             \tCFSR  0x{:x}\r\n\
-             \tHSFR  0x{:x}\r\n\
-             \tInstruction Access Violation:       {}\r\n\
-             \tData Access Violation:              {}\r\n\
-             \tMemory Management Unstacking Fault: {}\r\n\
-             \tMemory Management Stacking Fault:   {}\r\n\
-             \tMemory Management Lazy FP Fault:    {}\r\n\
-             \tInstruction Bus Error:              {}\r\n\
-             \tPrecise Data Bus Error:             {}\r\n\
-             \tImprecise Data Bus Error:           {}\r\n\
-             \tBus Unstacking Fault:               {}\r\n\
-             \tBus Stacking Fault:                 {}\r\n\
-             \tBus Lazy FP Fault:                  {}\r\n\
-             \tUndefined Instruction Usage Fault:  {}\r\n\
-             \tInvalid State Usage Fault:          {}\r\n\
-             \tInvalid PC Load Usage Fault:        {}\r\n\
-             \tNo Coprocessor Usage Fault:         {}\r\n\
-             \tUnaligned Access Usage Fault:       {}\r\n\
-             \tDivide By Zero:                     {}\r\n\
-             \tBus Fault on Vector Table Read:     {}\r\n\
-             \tForced Hard Fault:                  {}\r\n\
-             \tFaulting Memory Address: (valid: {}) {:#010X}\r\n\
-             \tBus Fault Address:       (valid: {}) {:#010X}\r\n\
-             ",
-            mode_str,
-            env!("TOCK_KERNEL_VERSION"),
-            stacked_r0,
-            stacked_r1,
-            stacked_r2,
-            stacked_r3,
-            stacked_r12,
-            stacked_lr,
-            stacked_pc,
-            stacked_xpsr,
-            (stacked_xpsr >> 31) & 0x1,
-            (stacked_xpsr >> 30) & 0x1,
-            (stacked_xpsr >> 29) & 0x1,
-            (stacked_xpsr >> 28) & 0x1,
-            (stacked_xpsr >> 27) & 0x1,
-            (stacked_xpsr >> 19) & 0x1,
-            (stacked_xpsr >> 18) & 0x1,
-            (stacked_xpsr >> 17) & 0x1,
-            (stacked_xpsr >> 16) & 0x1,
-            ici_it,
-            thumb_bit,
-            exception_number,
-            ipsr_isr_number_to_str(exception_number),
-            faulting_stack as u32,
-            (_estack as *const ()) as u32,
-            (&_ezero as *const u32) as u32,
-            shcsr,
-            cfsr,
-            hfsr,
-            iaccviol,
-            daccviol,
-            munstkerr,
-            mstkerr,
-            mlsperr,
-            ibuserr,
-            preciserr,
-            impreciserr,
-            unstkerr,
-            stkerr,
-            lsperr,
-            undefinstr,
-            invstate,
-            invpc,
-            nocp,
-            unaligned,
-            divbysero,
-            vecttbl,
-            forced,
-            mmfarvalid,
-            mmfar,
-            bfarvalid,
-            bfar
-        );
+        kernel_hardfault(faulting_stack);
```

Comment:
Was splitting this up into two functions important? Or is this just an artifact of debugging?

Reply:
"The fix is to mark them volatile as well as move the non-assembly code in hardfault_handler to a separate function so we can make hardfault_handler naked like the other ISRs." I think the non-assembly needs the prologue/epilogue functions but the raw assembly is just fine without them?

Reply:
Hmmm. It's not clear to me what prologue/epilogue the rest of the hardfault handler would need. Registers should already be stacked before entry into the handler.

Reply:
It's not directly related to the bug, because the issue with the hardfault handler only matters once we actually have a hard fault of some sort. But...

Without separating them, and being able to mark the hardfault handler `naked`, the prologue was pushing `LR` onto the stack, and the epilogue was popping directly into `PC`.

So, when an _app_ faulted, our changes of the `LR` register to return to the kernel instead of back to the process had no effect. Hardcoding a `BX LR` in the assembly isn't enough because at that point the kernel stack has changed (since the procedure _does_ push about 52 bytes of actual values on the stack), and the hardware unstacking is incorrect.

Separating it out in this way allows the us to mark the ISR as `naked`, thus ensuring the compiler to doesn't do any special handling of the LR register, but and all of the variable stacking/unstacking happens in the helper function, which is marked `inline(never)`.

Reply:
This all makes sense and sounds good, but I believe it means that #1176 is still a problem that needs addressing (independent of this PR)

Reply:
Correct, it does not fix #1176 (presumably it maybe somewhat isolates it, but is basically unrelated).

Reply:
I'm still missing something. Why can't you mark the function as `naked` but still keep the helper function code in the function as well?

Reply:
a `naked` function doesn't, e.g., stack callee saved registers (r4-r11), but the helper function code might (in this case definitely does) override those registers. You _could_ explicitly stack them in assembly with a custom prelude, but then you have to unstack them, and you don't necessarily have a guarantee at the end of your naked function what exactly the helper code has done with the stack.

So, maybe it's possible, but certainly simpler to reason about by moving the actual Rust code to a separate helper function.

Reply:
Oh! What I had missed here was that the NVIC doesn't stack r4-r11 for you, so you need to do it yourself.